src/mofaflex/_core/mofaflex.py

```diff
@@ -550,13 +585,72 @@ def _setup_gp(self, covariates=None, full_setup=True):
             self._gp_group_names = None
         return gp_warp_groups_order
 
-    def _setup_svi(self, prior_scales, init_tensor, covariates, feature_means, sample_means):
+    def _setup_guiding_vars(self, guiding_vars):
+        self._guiding_vars_names = (
+            list(self._data_opts.guiding_vars_obs_keys.keys()) if self._data_opts.guiding_vars_obs_keys else []
+        )
+        self._n_guiding_vars = len(self._guiding_vars_names)
+
+        if self._n_guiding_vars == 0:
+            self._guiding_vars_factors = {}
+            self._guiding_vars_categories = {}
+            return
+
+        # if no likelihood scale is provided, use the mean of the number of features
+        for guiding_var in self._guiding_vars_names:
+            if self._model_opts.guiding_vars_likelihood_scales[guiding_var] is None:
+                self._model_opts.guiding_vars_likelihood_scales[guiding_var] = np.array(
+                    list(self.n_features.values())
+                ).mean()
+
+        # update global number of factors
+        self._model_opts.n_factors = self._model_opts.n_factors + self._n_guiding_vars
+
+        # update global factor names (dense factors + guiding vars + informed factors)
+        self._factor_names = np.concatenate(
+            [
+                self._factor_names[: self.n_dense_factors],
+                self._guiding_vars_names,
+                self._factor_names[self.n_dense_factors + self._n_guiding_vars - 1 :],
+            ]
+        )
+
+        # create mapping from guiding var names to factor indices
+        self._guiding_vars_factors = {}
+        for i, guiding_var_name in enumerate(self._guiding_vars_names):
+            self._guiding_vars_factors[guiding_var_name] = self._n_informed_factors + i
+
+        # get unique categories for each guiding variable
+        self._guiding_vars_categories = {}
+        for guiding_var_name, guiding_var_likelihood in self._model_opts.guiding_vars_likelihoods.items():
+            if guiding_var_likelihood == "Categorical":
+                guiding_var_categories = []
+                # find unique categories for each group and then use union
+                for group_name in self.group_names:
+                    guiding_var_categories.append(
+                        np.unique(guiding_vars.datasets[guiding_var_name].covariates[group_name])
+                    )
+                self._guiding_vars_categories[guiding_var_name] = np.unique(np.concatenate(guiding_var_categories))
+            else:
+                # if not categorical, set to default
+                self._guiding_vars_categories[guiding_var_name] = np.array(["default"])
```

Comment:
If I'm following the code correctly, `guiding_vars.datasets[guiding_var_name].covariates[group_name])` will be integer codes for the categories, not strings. I don't think it makes sense to store those, probablly just keeping the number of categories for each guiding viariable is sufficient. The categories vectors are also only used to determine the number of categories in the model.